trie-db/src/node.rs

```diff
@@ -210,34 +212,59 @@ impl Node<'_> {
 			Self::Extension(n, h) =>
 				Ok(NodeOwned::Extension((*n).into(), h.to_owned_handle::<L>()?)),
 			Self::Branch(childs, data) => {
-				let mut childs_owned = [(); nibble_ops::NIBBLE_LENGTH].map(|_| None);
-				childs
-					.iter()
-					.enumerate()
-					.map(|(i, c)| {
-						childs_owned[i] =
-							c.as_ref().map(|c| c.to_owned_handle::<L>()).transpose()?;
-						Ok(())
-					})
-					.collect::<Result<_, _, _>>()?;
-
-				Ok(NodeOwned::Branch(childs_owned, data.as_ref().map(|d| d.to_owned_value::<L>())))
+				let childs_owned = if let Some(last_existing_child_index) =
+					childs.iter().rposition(Option::is_some)
+				{
+					let mut childs_owned = Vec::with_capacity(last_existing_child_index + 1);
+
+					childs
+						.iter()
+						.enumerate()
+						.map(|(i, c)| {
+							if i <= last_existing_child_index {
+								childs_owned.push(
+									c.as_ref().map(|c| c.to_owned_handle::<L>()).transpose()?,
+								);
+							}
+							Ok(())
+						})
+						.collect::<Result<_, _, _>>()?;
+					childs_owned
+				} else {
+					Vec::new()
+				};
+
+				Ok(NodeOwned::Branch(
+					ChildrenNodesOwned(childs_owned),
+					data.as_ref().map(|d| d.to_owned_value::<L>()),
+				))
 			},
 			Self::NibbledBranch(n, childs, data) => {
-				let mut childs_owned = [(); nibble_ops::NIBBLE_LENGTH].map(|_| None);
-				childs
-					.iter()
-					.enumerate()
-					.map(|(i, c)| {
-						childs_owned[i] =
-							c.as_ref().map(|c| c.to_owned_handle::<L>()).transpose()?;
-						Ok(())
-					})
-					.collect::<Result<_, _, _>>()?;
+				let childs_owned = if let Some(last_existing_child_index) =
+					childs.iter().rposition(Option::is_some)
+				{
+					let mut childs_owned = Vec::with_capacity(last_existing_child_index + 1);
+
+					childs
+						.iter()
+						.enumerate()
+						.map(|(i, c)| {
+							if i <= last_existing_child_index {
+								childs_owned.push(
+									c.as_ref().map(|c| c.to_owned_handle::<L>()).transpose()?,
+								);
+							}
+							Ok(())
+						})
+						.collect::<Result<_, _, _>>()?;
+					childs_owned
+				} else {
+					Vec::new()
```

Comment:
Should specify capacity explicitely here (no real reason to specify it when there is a node and not here). (same in previous code).

Reply:
You mean calling it Vec::with_capacity(0) ?

Reply:
yes or something else if consider better, just to not depend on default behavior of new